Add player

lrharris215 opened the pull request:
A human player can now place 1 disc on the game board. The disc will appear in the column of their choice. I also upgraded Gradle and configured the program to work in the terminal .

lrharris215 committed:
Fix Board#addDisc

Board#addDisc now only accepts a disc and a column number
as arguments, and places the disc at the bottom of the
column. Also adds Board#isEmpty to check if the space is
occupied or not.

lrharris215 committed:
Add Player and PlayerScanner Classes

The Player class takes in a disc, and a PlayerScanner. The
PlayerScanner takes in a Scanner, and uses it to receive
input from the terminal. The player can take their turn
by placing a disc in the inputted column of the board.

lrharris215 committed:
Test: Player#takeTurn

Also made Player#placeDisc private. The test checks that
the disc is placed in the correct column. Uses a TestDisc
and TestScanner class.

lrharris215 committed:
Add Scannable Interface

The Scannable interface simplifies the structure of my
Player tests. Allows my TestScanner to exist without
extending PlayerScanner.

lrharris215 committed:
Add Player to ConnectFour

Add a player instance to the ConnectFour main method. Player
can now input a move and see it appear on the screen.

lrharris215 committed:
Refactor: Add setUpGame

Moved all the object initialization into a separate
function.

lrharris215 committed:
Add some more text to terminal instructions

Adds instructions for the player to select a column. Also
added numbers to the columns. Player now has a name.

lrharris215 committed:
Refactor Player#takeTurn

Converts choice to choice - 1 for indexing.

lrharris215 committed:
Fix: Gradle bug

./gradlew run was breaking before because of the default
input. Changing it to System.in in the build file fixed it.

lrharris215 committed:
Fix: Upgraded Gradle Version

Trying to fix the deprecation notice broke everything else.
Turned out that my version of Gradle was not compatible with
Java 17. Upgrading to Gradle 7.3 fixed it, and now
<./gradlew run> works without long messages.

lrharris215 committed:
Merge branch 'main' into addPlayer

kudukacar commented on src/main/java/com/connectfour/ConnectFour.java:
```diff
@@ -1,18 +1,43 @@
+        nextTurn(playerOne);
+
         Presenter.printBoard(display, board);
     }
```
You could abstract this out even more.  Main could set up the game and then call a play or run method in another class to start the game.  That class would have the steps to play the game - welcome, present board, take turn etc. You can also write some tests for that class.

lrharris215 replied:
Yeah I wasn't sure if that was over-coding at this step, since the game isn't looping yet. But I can do that!

pumpkincouture commented on build.gradle:
```diff
@@ -20,5 +20,9 @@ test {
 
 application {
-    mainClassName = 'com.connectfour.ConnectFour'
+    getMainClass().set('com.connectfour.ConnectFour')
```
What's the advantage of doing this over what you had before?

[EDIT: I assume it's a Gradle thing!)

lrharris215 replied:
Gradle was giving a notice every time it ran that this was deprecated. Apparently this is the right way to do it now.

lrharris215 committed:
Refactor: Add Game class

The Game class now handles the actual running of the game,
while ConnectFour is in charge of setting up the components
needed. Game has start(), play(), and nextTurn(player) functions.

pumpkincouture commented on src/main/java/com/connectfour/ConnectFour.java:
```diff
@@ -1,18 +1,43 @@
+
+    private static void nextTurn(Player player){
+        display.println(player.name + ", please select which column you would like to place your disc.");
+        display.println("Your color is " + player.disc.getColor());
```
Let's store the strings as static variables in this class! (since they won't be used anywhere else)

lrharris215 replied:
Kk, will do!

lrharris215 committed:
Test: Add Board#addDisc tests

Added tests for the addDisc function. Also created consts for
the strings used in Game.

lrharris215 committed:
Refactor: Moved static vars to local vars

lrharris215 committed:
Refactor: Add Playerable interface

This will make it easier to test Game. Enables creation of
TestPlayer for Game tests.

lrharris215 committed:
Test: Add Game tests

Add test for Game#start, checks for welcome message

lrharris215 committed:
Refactor: Moved test classes to own file

Moved all the Test classes to their own file to keep code DRY

lrharris215 committed:
Fix: TestDisc bug

Changed color of TestDisc to something not used in program

lrharris215 committed:
Fix: TestDisplay states

Changed TestDisplay states to ArrayLists

lrharris215 committed:
Add: Test for Game#play

kudukacar reviewed (approved):
Looks good!
